api/controller.py

```diff
@@ -619,7 +654,62 @@ def borrow(self, data_source, identifier_type, identifier, mechanism_id=None):
         headers = { "Content-Type" : OPDSFeed.ACQUISITION_FEED_TYPE }
         return Response(content, status_code, headers)
 
-    def fulfill(self, data_source, identifier_type, identifier, mechanism_id=None, do_get=None):
+    def best_lendable_pool(self, library, patron, identifier_type, identifier,
+                           mechanism_id):
+        """Of the available LicensePools for the given Identifier, return the
+        one that's the best candidate for loaning out right now.
+        """
+        # Turn source + identifier into a LicensePool
+        pools = self.load_licensepools(
+            self.library, identifier_type, identifier
+        )
+        if isinstance(pools, ProblemDetail):
+            # Something went wrong.
+            return pools
+
+        best = None
+        mechanism = None
+        problem_doc = None
+        
+        # We found a number of LicensePools. Try to locate one that
+        # we can actually loan to the patron.
+        for pool in pools:
```

Comment:
Since we're iterating over all the pools, this would be a good place to check whether the patron already has this book out on loan (as per my previous comment). It wouldn't be too hard to take care of the TODO within this branch.